File: leetcode/maxDepth.py
# 二叉树高度 非递归
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def maxDepth(self, root: TreeNode) -> int:
        if root == None:
            return 0
        head = [root]
        i = 0
        while head:
            next_layer =[]
            i += 1
            for node in head:
                if node.left == None and node.right != None:
                    next_layer.append(node.right)
                if node.right == None and node.left != None:
                    next_layer.append(node.left)
                if node.left != None and node.right != None:
                    next_layer.extend([node.left,node.right])
            head = next_layer
        return i
#递归版本
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def maxDepth(self, root: TreeNode) -> int:
        if root == None:
            return 0
        else:
            leftDepth = self.maxDepth(root.left)
            rightDepth = self.maxDepth(root.right)
            return max(leftDepth,rightDepth) + 1
        
        

File: Leetcode_Java/maxDepth.java
#非递归

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
import java.util.*;

class Solution {
    public int maxDepth(TreeNode root) {
        if (root == null){
            return 0;
        }
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        int preSum = 0;
        while (!queue.isEmpty()){
            int count = queue.size();
            while (count > 0) {
                TreeNode node = queue.poll();
                if (node.left != null) {
                    queue.add(node.left);
                }
                if (node.right != null){
                    queue.add(node.right);
                }
                count --;
            }
            preSum ++;
        }
        return preSum;
    }
}

#递归

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int maxDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        else
        {
            int leftHeight = maxDepth(root.left);
            int rightHeight = maxDepth(root.right);
            return java.lang.Math.max(leftHeight,rightHeight) + 1;       
        }
    }
}
